pkg/logproto/indexgateway.proto

```diff
@@ -25,4 +26,45 @@ service IndexGateway {
   // Note: this MUST be the same as the variant defined in
   // logproto.proto on the Querier service.
   rpc GetVolume(logproto.VolumeRequest) returns (logproto.VolumeResponse) {}
+
+  // GetShards is an optimized implemented shard-planning implementation
+  // on the index gateway and not on the ingester.
+  rpc GetShards(ShardsRequest) returns (stream ShardsResponse);
+}
+
+message ShardsRequest {
+  int64 from = 1 [
+    (gogoproto.customtype) = "github.com/prometheus/common/model.Time",
+    (gogoproto.nullable) = false
+  ];
+  int64 through = 2 [
+    (gogoproto.customtype) = "github.com/prometheus/common/model.Time",
+    (gogoproto.nullable) = false
+  ];
+  string query = 3;
```

Comment:
I think when we create this ShardRequest in the QF, we have already parsed the plan form the query. maybe we can pass the Plan here instead of the raw query? This way in the future, GetShards can take advantage of whichever information the plan contains.